RequireJS optimization

smagch opened the pull request:
I think libraries like `jquery`, `underscore` and  `backbone` would be maintainable with `require dependencies` though it makes config relatively large.

I have no idea whether thimble do conditional loading or not.
If it don't do dynamic conditional loading, require.js would be replaced with [almond.js](https://github.com/jrburke/almond).

This pull request have some points which should be improve: duplicate config in `index.html`, `test/index.html`, `build-require.js`... :(
I'm making config a JSON file and inserting it during server-side templating in my personal backbone project.
I haven't looked into webpagemaker, so you may know better solutions.

toolness/friendlycode#112

smagch committed:
update requireJS v2.0.2

smagch committed:
make jquery, underscore and backbone require dependencies

smagch committed:
update require config

smagch committed:
make template plugin optimizable

smagch committed:
fix duplicate appReady

smagch committed:
add main require for optimization

smagch committed:
add build script

smagch committed:
add main-build comment out

toolness commented:
Very cool, thanks for all this work @smagch! Friendlycode doesn't currently do conditional loading, with the possible exception of TypeKit, so RequireJS should _probably_ be fine. I'll take a look at your stuff today, sorry for the really long delay on this.

toolness committed:
added jquery as dependency for test-app.js

smagch commented:
> There isn't a Python variant of this

`r.js` also [works with Rhino](https://github.com/jrburke/r.js/#java) though I haven't tried that: I haven't set up `.jar` file by myself because Eclipse do everything when I was working as a Java developer.

It would be nice not to merge build script I wrote because it has ugly copy and paste of config file :(

I'm using Makefile and node build script to build my app which use `require.js` and `backbone.js` in my personal project.
https://github.com/dogazei/dogazei.org/blob/v0.1.2/Makefile#L67-83
https://github.com/dogazei/dogazei.org/blob/v0.1.2/support/buildRequire.js
During development, I just simply comment out optimized script and use `main.js`
https://github.com/dogazei/dogazei.org/blob/v0.1.2/views/holy-potato/index.jade#L18-22
Before deployment, I build scripts with Makefile and comment out `main.js` and use optimized script.

When you install r.js via npm, it seems to be able to optimize by command line! But I haven't figure out how to use that correctly: I ended up with error.

toolness committed:
Consolidated duplicate requirejs config into js/require-config.js.

This file can be loaded both as a browser script *BEFORE* requirejs, and
as a module from node.

toolness committed:
Added .gitignore.

toolness committed:
Callback to requirejs.optimize() is list of output modules, not err boolean.

toolness committed:
Use _.noConflict() and jQuery.noConflict() in their shims.

This will help ensure that the global namespace is clean and
that no modules accidentally get away with using a global instead
of an injected dependency (which is particularly dangerous now,
since it introduces race conditions).

toolness commented:
I've just offered a few fixes to this here: https://github.com/smagch/friendlycode/pull/1. Let me know what you think.

toolness commented:
Hmm, found one bug w/ the pull request, I think, which is that error messages don't seem to work anymore. If I create an invalid tag like `<pe>`, for instance, no error message appears and this error gets logged to the console:

```
Error template not found for INVALID_TAG_NAME @ http://localhost:8001/slowparse/spec/errors.jquery.js:91
```

The error templates are supposed to be loaded as part of the `appReady` requirejs plugin, I think.

smagch commented:
Ah, I didn't notice the bug! It's weird... help messages are shown correctly.

smagch commented on js/template.js:
```diff
@@ -1,8 +1,32 @@
+      if (buildMap[moduleName]) {
+        var content = buildMap[moduleName];
+        write.asModule(pluginName + "!" + moduleName,
+          "define(function () { \n  return " + content + ";});\n");
```
I missed dependencies. It need to be

``` javascript
"define(['underscore'], function (_) { \n  return " + content + ";});\n");
```

smagch replied:
49a0d34

toolness commented:
Hmm, something else strange seems to be going on--it's as though two copies of `jQuery` are in the page at once. I'm getting an error that `$.fn.fillError()` doesn't exist in `fc/ui/error-help.js`, even though the file it's defined in (`slowparse/spec/errors.jquery.js`) has been loaded beforehand.

smagch commented:
> (`slowparse/spec/errors.jquery.js` ) has been loaded beforehand.

[error-help.js](https://github.com/toolness/friendlycode/blob/smagch-optimize/js/fc/ui/error-help.js#L5) currently don't have `jquery-slowparse` dependencies. So it could be possible to error, couldn't it?

smagch committed:
Merge pull request #1 from toolness/smagch-optimize

Fixed some issues w/ requirejs integration

smagch committed:
add underscore dependencies for optimization

smagch committed:
load appReady as plugin.
As long as I see official documents, it is standard to use "req" inside load method
rather than direct dependencies.
http://requirejs.org/docs/plugins.html#apiload

smagch committed:
add jquery-slowparse dependency instead of jquery since there is $.fn.fillError call

toolness commented:
> [error-help.js](https://github.com/toolness/friendlycode/blob/smagch-optimize/js/fc/ui/error-help.js#L5) currently don't have `jquery-slowparse` dependencies. So it could be possible to error, couldn't it?

Well, the exception coming out of error-help is raised a long time after initial page load, when I modify a document to contain an error... I think the missed dependency would be a good explanation if the exception were happening during load time, but since the dependency is just a shim, the jquery-slowparse plugin should have modified the global `jQuery` object by the time I've typed an error into the editor.

toolness commented:
Huh, something you did fixed the problem. Everything looks good now, so I'm going to merge this pull request. Thanks for everything!

toolness commented:
It looks like the multiple-instances-of-jQuery problem I was observing was due to tabzilla trying to load jQuery itself! Fixed this in dfbd4c21335868f3f1e30f8a35852af6e248107a.

smagch commented:
I could add `include: "require.min"` in [build config](https://github.com/toolness/friendlycode/blob/dfbd4c21335868f3f1e30f8a35852af6e248107a/build-require.js#L21-31).
It'll concatenate `require.js` and optimized main script.

Instead of 

``` html
<script src="js/require.min.js" data-main="js/main-built.js"></script>
```

It'll be simply like this.

``` html
<script src="js/main-built.js"></script>
```

I just downloaded [almond.js](https://github.com/jrburke/almond) and tried `include: "almond"`. But it didn't work out: it seems `require("typekit-ready!")` have a trouble.
Anyway, I'll let you know if I have time and figure out a reason of the error.

Thanks, I've learned a lot through the pull request.
